### PR TITLE
[CVAT integration] Use pixelwise masks, not polygons, for instance segmentation #4483

### PR DESCRIPTION
**Pull request by NicDionne**
The following items are still pending:

- [x] Code to upload fiftyone to CVAT pixel mask
- [x] Unit test for uploading to CVAT
- [x] Unit test for converting CVAT shapes back into the FiftyOne format
- [x] Implementation for converting CVAT shapes back into FiftyOne format

@ehofesmann  
1. Sorry, I'm not sure how I should go about testing CVAT components with API request ?
2. I'm also unsure how to incorporate the download from CVAT to Fiftyone. I attempted to use load_annotation_results, but it didn't seem to trigger this particular part of the code. Could you advise on how to proceed?https://github.com/voxel51/fiftyone/blob/77f664c6da47ce05abb36326237fc10aef5fe999/fiftyone/utils/cvat.py#L5908

## What changes are proposed in this pull request?

Propose following the issue #4483, to make it possible to export fiftyone mask to CVAT as pixel mask.

## How is this patch tested? If it is not, please explain why.

For now, simple upload test with rectangle. I seek guidance regarding how it should be properly tested. **For now this pull request is only a draft.**

### Is this a user-facing change that should be mentioned in the release notes?

Yes, 
> ```python
> dataset.annotate(anno_key, label_field='detections', label_type='instances')
> ```
The data should now be uploaded in pixel format instead of polygons. This change is good because CVAT does not support polygons with holes, which could create bottlenecks in certain use cases.

-   [ ] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [x] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ ] Other

## Summary by CodeRabbit

- **New Features**
	- Introduced a new class for converting between run-length encoded masks and binary image masks.
	- Added methods to facilitate the conversion of masks for better integration with the FiftyOne framework.
	- Enhanced shape conversion capabilities to support instance detection with masks.
- **Bug Fixes**
	- Updated the export process to properly handle new mask types during annotation uploads.

### CODE REVIEW

**Review by ehofesmann — COMMENTED** [review]
This is awesome! Thank you so much for taking this on @NicDionne !!

To your questions:

1. We don't have a great way to mock the CVAT API at the moment, so most testing is done against a running CVAT instance. In the end, it would be great to add a new test for this to our test suite [here](https://github.com/voxel51/fiftyone/blob/77f664c6da47ce05abb36326237fc10aef5fe999/tests/intensive/cvat_tests.py).
2. In order to be able to load the CVAT mask back into FiftyOne, you'd need to add a new check [here](https://github.com/voxel51/fiftyone/blob/77f664c6da47ce05abb36326237fc10aef5fe999/fiftyone/utils/cvat.py#L5895) for `if shape_type == "mask"`, which can then check for `instance` or `instances` types and call your new `CVATShape.to_instance_detection()` method. Let me know if you run into any issues with this!

**Comment by NicDionne** [comment]
Well there seems to be a problem on my side to commit files from the cvat_test.py. Pylint is going crazy about E1101 errors.

Here is the code for the unit test I ran.

File: fiftyone/utils/cvat.py [hunk 1]
```diff
@@ -7076,6 +7104,38 @@ def to_detection(self):
...
+        """
+        xtl, ytl, xbr, ybr = self.points[-4:]
+        rel = np.array(self.points[:-4], dtype=int)
+        frame_width, frame_height = self.frame_size
+        mask_w, mask_h = (
+            round(xbr - xtl) + 1,
+            round(ybr - ytl) + 1,
+        )  # We need to add 1 because cvat uses - 1
+        mask = HasCVATBinMask.rle_to_binary_image_mask(
+            rel, mask_width=mask_h, mask_height=mask_w
```
[the remark this answers is not in the document]
Comment (NicDionne) [reply fiftyone/utils/cvat.py]: I pushed a fix in commit 513b633, please review it.

**Comment by NicDionne** [comment]
Hi @brimoor, 

It looks like the build failed, but strangely it stopped at FFmpeg. I'm not really sure what I might have touched that could cause this. Do you have any ideas?

Thanks,

**Comment by NicDionne** [comment]
Hi @ehofesmann quick remind on the PR :)

**Comment by NicDionne** [comment]
Hi @ehofesmann @brimoor quick remind on the PR :)

**Comment by ehofesmann** [comment]
Hey @NicDionne @brimoor
Yes! Sorry, I got wrapped up in some things I will give this another pass soon. The ffmpeg build errors are something that we are aware of and not a concern for this PR

**Comment by NicDionne** [comment]
Hi @ehofesmann  @brimoor,

Quick remind on the PR....

**Comment by NicDionne** [comment]
Hi @ehofesmann @brimoor,

Quick remind on the PR....

**Review by brimoor — APPROVED** [review]
@NicDionne apologies for the delay here, I'll get this across the finish line now!

**Comment by NicDionne** [comment]
Thanks !

**Pull request merged**